[lldb][debugserver] Read/write SME registers on arm64

jasonmolenda opened the pull request:
**Note:** The register reading and writing depends on new register flavor support in thread_get_state/thread_set_state in the kernel, which will be first available in macOS 15.4.

The Apple M4 line of cores includes the Scalable Matrix Extension (SME) feature. The M4s do not implement Scalable Vector Extension (SVE), although the processor is in Streaming SVE Mode when the SME is being used.  The most obvious side effects of being in SSVE Mode are that (on the M4 cores) NEON instructions cannot be used, and watchpoints may get false positives, the address comparisons are done at a lowered granularity.

When SSVE mode is enabled, the kernel will provide the Streaming Vector Length register, which is a maximum of 64 bytes with the M4. Also provided are SVCR (with bits indicating if SSVE mode and SME mode are enabled), TPIDR2, SVL.  Then the SVE registers Z0..31 (SVL bytes long), P0..15 (SVL/8 bytes), the ZA matrix register (SVL*SVL bytes), and the M4 supports SME2, so the ZT0 register (64 bytes).

When SSVE/SME are disabled, none of these registers are provided by the kernel - reads and writes of them will fail.

Unlike Linux, lldb cannot modify the SVL through a thread_set_state call, or change the processor state's SSVE/SME status.  There is also no way for a process to request a lowered SVL size today, so the work that David did to handle VL/SVL changing while stepping through a process is not an issue on Darwin today.  But debugserver should be providing everything necessary so we can reuse all of David's work on resizing the register contexts in lldb if it happens in the future.  debugbserver sends svl, svcr, and tpidr2 in the expedited registers when a thread stops, if SSVE|SME mode are enabled (if the kernel allows it to read the ARM_SME_STATE register set).

While the maximum SVL is 64 bytes on M4, the AArch64 maximum possible SVL is 256; this would give us a 64k ZA register.  If debugserver sized all of its register contexts assuming the largest possible SVL, we could easily use 2MB more memory for the register contexts of all threads in a process -- and on iOS et al, processes must run within a small memory allotment and this would push us over that.

Much of the work in debugserver was changing the arm64 register context from being a static compile-time array of register sets, to being initialized at runtime if debugserver is running on a machine with SME.  The ZA is only created to the machine's actual maximum SVL. The size of the 32 SVE Z registers is less significant so I am statically allocating those to the architecturally largest possible SVL value today.

Also, debugserver includes information about registers that share the same part of the register file.  e.g. S0 and D0 are the lower parts of the NEON 128-bit V0 register.  And when running on an SME machine, v0 is the lower 128 bits of the SVE Z0 register.  So the register maps used when defining the VFP registers must differ depending on the capabilities of the cpu at runtime.

I also changed register reading in debugserver, where formerly when debugserver was asked to read a register, and the thread_get_state read of that register failed, it would return all zero's.  This is necessary when constructing a `g` packet that gets all registers - because there is no separation between register bytes, the offsets are fixed.  But when we are asking for a single register (e.g.  Z0) when not in SSVE/SME mode, this should return an error.

This does mean that when you're running on an SME capabable machine, but not in SME mode, and do `register read -a`, lldb will report that 48 SVE registers were unavailable and 5 SME registers were unavailable.  But that's only when `-a` is used.

The register reading and writing depends on new register flavor support in thread_get_state/thread_set_state in the kernel, which is not yet in a release.  The test case I wrote is skipped on current OSes.  I pilfered the SME register setup from some of David's existing SME test files; there were a few Linux specific details in those tests that they weren't easy to reuse on Darwin.

rdar://121608074

jasonmolenda committed:
[lldb][debugserver] Read/write SME registers on arm64

The Apple M4 line of cores includes the Scalable Matrix Extension
(SME) feature. The M4s do not implement Scalable Vector Extension
(SVE), although the processor is in Streaming SVE Mode when the SME
is being used.  The most obvious side effects of being in SSVE Mode
are that (on the M4 cores) NEON instructions cannot be used, and
watchpoints may get false positives, the address comparisons are
done at a lowered granularity.

When SSVE mode is enabled, the kernel will provide the Streaming
Vector Length register, which is a maximum of 64 bytes with the M4.
Also provided are SVCR (with bits indicating if SSVE mode and SME
mode are enabled), TPIDR2, SVL.  Then the SVE registers Z0..31 (SVL
bytes long), P0..15 (SVL/8 bytes), the ZA matrix register (SVL*SVL
bytes), and the M4 supports SME2, so the ZT0 register (64 bytes).

When SSVE/SME are disabled, none of these registers are provided by
the kernel - reads and writes of them will fail.

Unlike Linux, lldb cannot modify the SVL through a thread_set_state
call, or change the processor state's SSVE/SME status.  There is
also no way for a process to request a lowered SVL size today, so
the work that David did to handle VL/SVL changing while stepping
through a process is not an issue on Darwin today.  But debugserver
should be providing everything necessary so we can reuse all of
David's work on resizing the register contexts in lldb if it happens
in the future.  debugbserver sends svl, svcr, and tpidr2 in the
expedited registers when a thread stops, if SSVE|SME mode are enabled
(if the kernel allows it to read the ARM_SME_STATE register set).

While the maximum SVL is 64 bytes on M4, the AArch64 maximum possible
SVL is 256; this would give us a 65k ZA register.  If debugserver
sized all of its register contexts assuming the largest possible
SVL, we could easily use 2MB more memory for the register contexts
of all threads in a process -- and on iOS et al, processes must run
within a small memory allotment and this would push us over that.

Much of the work in debugserver was changing the arm64 register
context from being a static compile-time array of register sets,
to being initialized at runtime if debugserver is running on a
machine with SME.  The ZA is only created to the machine's actual
maximum SVL. The size of the 32 SVE Z registers is less significant
so I am statically allocating those to the architecturally largest
possible SVL value today.

Also, debugserver includes information about registers that share
the same part of the register file.  e.g. S0 and D0 are the lower
parts of the NEON 128-bit V0 register.  And when running on an SME
machine, v0 is the lower 128 bits of the SVE Z0 register.  So the
register maps used when defining the VFP registers must differ
depending on the runtime state of the cpu.

I also changed register reading in debugserver, where formerly when
debugserver was asked to read a register, and the thread_get_state
read of that register failed, it would return all zero's.  This is
necessary when constructing a `g` packet that gets all registers -
because there is no separation between register bytes, the offsets
are fixed.  But when we are asking for a single register (e.g.  Z0)
when not in SSVE/SME mode, this should return an error.

This does mean that when you're running on an SME capabable machine,
but not in SME mode, and do `register read -a`, lldb will report
that 48 SVE registers were unavailable and 5 SME registers were
unavailable.  But that's only when `-a` is used.

The register reading and writing depends on new register flavor
support in thread_get_state/thread_set_state in the kernel, which
is not yet in a release.  The test case I wrote is skipped on current
OSes.  I pilfered the SME register setup from some of David's
existing SME test files; there were a few Linux specific details
in those tests that they weren't easy to reuse on Darwin.

rdar://121608074

jasonmolenda commented:
One difference from debugserver and lldb-server is that lldb-server provides "vg" and "svg" registers (vector granule, streaming vector granule, depending on Streaming mode) which is the vector length in 8-byte granules.  On Darwin, debugserver provides only "svl", in bytes.  I considered having debugserver report the vector length in granules to match the Linux behavior, but the kernel was giving me the value in bytes and I think it's a more natural representation, so I stuck with it.

jasonmolenda committed:
reformats refomatted

jasonmolenda committed:
reformat until refomatting bot is appeased

DavidSpickett commented:
> But when we are asking for a single register (e.g. Z0) when not in SSVE/SME mode, this should return an error.

I assume you did what we did for Linux, where the inactive registers do not dissapear from the register list but just fail to read.

Linux also returns all 0s for the array register when it's disabled, but this is I think more about keeping the register offsets valid than anything practical.

> I considered having debugserver report the vector length in granules to match the Linux behavior, but the kernel was giving me the value in bytes and I think it's a more natural representation, so I stuck with it.

To repeat what I said over email to Jason - 

This "vector granule" possibly comes from a time before SVE was finalised and we (Arm) did not know exactly how the vector length would be reported. Our simulators use this "granule" term and it made it into GDB via that.

The Linux kernel does not report in granules it does the same as Darwin, bytes. We (Linaro) added "vg" to be compatible with GDB and QEMU.

Bytes is the more useful reporting unit because the SVE programming question is always "how many units of X bytes fit into the vector length".

DavidSpickett commented on lldb/tools/debugserver/source/MacOSX/arm64/DNBArchImplARM64.cpp:
```diff
@@ -2387,6 +2863,36 @@ bool DNBArchMachARM64::SetRegisterValue(uint32_t set, uint32_t reg,
+      }
+      if (reg >= sve_p0 && reg <= sve_p15) {
+        memcpy(&m_state.context.sve.p[reg - sve_p0], &value->value.v_uint8,
+               256 / 8);
```
Oh, the max P register size?

If so please add a comment to say that.

jasonmolenda replied:
I should have used the current svl value / 8 here, like I did in other memcpy's, instead of the maximum possible P size.  thanks for catching it.

DavidSpickett reviewed (commented):
Lightly skimmed the debugserver parts, I assume one of your colleagues will help there.

jasonmolenda commented on lldb/test/API/macosx/sme-registers/main.c:
```diff
@@ -0,0 +1,111 @@
+
+  int c = 10; // break here
+  c += 5;
+  c += 5;
```
Yeah, I wanted to do a `stepi`, so I needed some non-function-call instructions once the SME state was set up.  clang likes to use NEON instructions when zeroing out objects, so I couldn't throw in a function call to libc without disabling SME first, or those instructions would fail on an M4.

DavidSpickett replied:
Ok right I didn't think about that.

Does this mean that executing expressions in streaming mode is patchy at best? We're reliant on Clang not to choose the wrong thing.

jasonmolenda replied:
Yeah I think you can disable streaming SVE mode (PSTATE.SM) and leave SME (PSTATE.ZA) enabled (so the ZA/ZT0 registers keep their current contents), to call a function or use FPSIMD instructions (and assuming the function you call doesn't do any SME things).  With the caveat that changing PSTATE.SM zeros out the Z/V and P registers.

jasonmolenda replied:
Ah, I just found the optional feature that M4 doesn't implement - FEAT_SME_FA64.  This would allow the use of SIMD instructions while in streaming SVE mode.

DavidSpickett replied:
LLDB should be able to recover from a SIGILL, so it will be surprising but at least they can continue the program.

jasonmolenda commented on lldb/test/API/macosx/sme-registers/TestSMERegistersDarwin.py:
```diff
@@ -0,0 +1,163 @@
+        zt0_final = zt0.GetChildAtIndex(63)
+        self.assertEqual(zt0_final.GetValueAsUnsigned(), 63)
+
+        z0_old_values = []
```
Ahhhhh I actually meant to go back to this, and try it clearer.  I forgot.  I'll do the minor improvement I had in mind.

JDevlieghere replied:
Not sure if this code changed since David's comment, but there's a lot of code duplication here. Seems like a little helper function could eliminate most of it without hurting debuggability?

jasonmolenda commented on lldb/test/API/macosx/sme-registers/TestSMERegistersDarwin.py:
```diff
@@ -0,0 +1,163 @@
+        for i in range(zt0.GetNumChildren()):
+            self.assertEqual(
+                zt0_old_values[i] + 2, zt0.GetChildAtIndex(i).GetValueAsUnsigned()
+            )
```
Good idea.

jasonmolenda commented on lldb/tools/debugserver/source/DNBDefs.h:
```diff
@@ -312,16 +312,21 @@ struct DNBRegisterValue {
+    // large enough to hold that much data.  The current Apple cores
+    // have a much smaller maximum ZA reg size, but there are not
+    // multiple copies of this object so increase the static size to
+    // maximum possible.
```
Yeah, this object is allocated to read/write a single register, so a read of x0 will be a 64k object.  But looking at the debugserver sources, we don't store an array of them anywhere - we read / write individual registers one at a time with this object for a short time period, so I don't think the memory increase is a problem.  It might be better to have a dynamically allocated size here though, as you did.  I did that for the DNBArm64ArchImpl register contexts stored for each thread, where we will have one for each thread when stopped, that memory use made me more nervous.

DavidSpickett replied:
I think for Linux we were also stack allocating the register value and I didn't want 64k stack frames everywhere we used one. https://github.com/llvm/llvm-project/commit/d99d9d8b74ffcb8ac418fcdafa750b07f2bd9935 in case any of the concerns apply to debugserver also.

(I am also very aware of these issues because in a previous job when we added MIPS MSA support we accidentally turned every register object into 512 bits, even the 8 and 16 bit ones we read from non-MIPS DSP chips)

jasonmolenda replied:
I looked at the `DNBRegisterValue` use a little more, and I think I want to change it to a heap allocated object, but it's going to touch all of the arch plugins in debugserver, so I will do it as a separate change from this one.  On the macOS environment, the single 64k register on the stack isn't blowing anything, but it's not ideal and could cause a problem in our more constrained environments.

JDevlieghere replied:
+1 on changing this to a heap object. This seems unnecessarily wasteful when not in SME mode, which I expect to remain the majority of the time.

jasonmolenda replied:
I was thinking of restructuring the internals of the object to heap-allocate the value space, which would require touching all of the DNBArchImpl back-ends, but actually just heap-allocating the object in RNBRemote (the main place this object is created) would be much easier than changing it at all.

jasonmolenda commented on lldb/tools/debugserver/source/MacOSX/arm64/DNBArchImplARM64.cpp:
```diff
@@ -93,6 +93,55 @@ DNBArchMachARM64::SoftwareBreakpointOpcode(nub_size_t byte_size) {
+// It will return the maximum SVL length for this process.
+uint16_t __attribute__((target("sme"))) DNBArchMachARM64::get_svl_bytes(void) {
+  uint64_t ret = 0;
+  asm volatile("rdsvl	%[ret], #1" : [ret] "=r"(ret));
```
Yes, when I added this there wasn't a way to get the maximum SVL but the kernel folks are adding `hw.optional.arm.sme_max_svl_b` sysctl.  You could imagine a scenario where we have generations of cores with a 64 byte SVL max, and then there's a new core some day with 128 byte SVL and there are programs which are hardcoded to the old length; maybe a process would be launched with a 64 byte SVL max by default.  It's possible to imagine debugserver having a lower max SVL than the inferior process which might have been launched with a "give me the actual full SVL" mode.  I should probably remove this backup method of getting the max SVL.

DavidSpickett replied:
That would be my concern, older debug server - newer CPU with a wider SME unit.

You could leave it in as a last resort but I wonder if it would even be understandable for users. If the kernel returns the registers in a buffer end to end, and they're 128 byte but debugserver thinks they're 64 byte. It'll likely 1. fail to read at all or 2. Show the first ZA row as the first two ZA rows.

As opposed to what might be useful in some very niche scenario, where you could see the bottom few bytes of each register.

Not that SME/SVE in its raw form is that useful to begin with.

jasonmolenda replied:
I had to install a newer kernel with the newly added sysctl to test & confirm that the sysctl works.  I'll remove the "read debugserver's max SVL and use it" fallback.

jasonmolenda commented on lldb/test/API/macosx/sme-registers/TestSMERegistersDarwin.py:
```diff
@@ -0,0 +1,163 @@
+            self.runCmd("reg read -a")
+
+        z0 = frame.register["z0"]
+        for i in range(z0.GetNumChildren()):
```
There is an `SBValue.children` property in Python, but this code needs to know the index number so it can read the old value array.

jasonmolenda committed:
Update test case:
 1. confirm that we cannot read SME/SVE regs
    when not in SSVE mode.
 2. Make it clearer how I'm modifying all of the
    SVE/SME registers, then instruction stepping, then
    reading them back to confirm that they were modified.

Remove `DNBArchMachARM64::get_svl_bytes`, depend entirely on the
hw.optional.arm.sme_max_svl_b sysctl to get the system's maximum
SVL, instead of debugserver's maximum SVL.  They're always the same
today, but it's possible to imagine it not being like that in the
future.

jasonmolenda commented:
Updated the API test case as per David's suggestions, and remove the code that was using debugserver's SVL as the hardware maximum, depending entirely on the newer sysctl to get the correct value instead.

jasonmolenda committed:
typeo fix

jasonmolenda committed:
clear up old/new values a bit more

DavidSpickett commented on lldb/tools/debugserver/source/RNBRemote.cpp:
```diff
@@ -2842,31 +2850,32 @@ rnb_err_t RNBRemote::SendStopReplyPacketForThread(nub_thread_t tid) {
+    for (uint32_t reg = 0; reg < g_num_reg_entries; reg++) {
+      int regset = g_reg_entries[reg].nub_info.set;
+      bool include_reg = false;
+      // Expedite interesting register sets, all registers not
```
Does this imply you are expediting svcr and tpidr2 as well?

I don't know debugserver but for Linux, only svl (svg) is needed in this scenario. Doesn't hurt to include more though.

DavidSpickett replied:
And by expedited I mean sent with every stop notification. If this code here is just what's sent with a general register read then it makes more sense to me.

jasonmolenda replied:
Yes, I figured as long as we were fetching ARM_SME_STATE register flavor to get svl, I might as well send svcr and tpidr2 as well, even if they're not being used today.

DavidSpickett replied:
You got me wondering how Linux tells whether it's in streaming mode or not. We rely on the fact that `vg` is the current vector length regardless of mode. So to update z0-31, we always use `vg`, and for `za` we use `svg` if we have it.

I think we would provide `vg` on a system with only streaming SVE, but I haven't tested it.

Given that, I think you expediting `svcr` is very sensible. If you have to deal with a bug along those lines, you won't have to guess what the mode is.

JDevlieghere commented on lldb/test/API/macosx/sme-registers/main.c:
```diff
@@ -0,0 +1,113 @@
+///  BUILT with
```
Nit: `s/built/build/`. But also this is covered by the makefile so maybe something like "Requires -mcpu=apple-m4" would be more to the point.

JDevlieghere commented on lldb/tools/debugserver/source/MacOSX/arm64/sme_thread_status.h:
```diff
@@ -0,0 +1,77 @@
+#include <mach/mach.h>
+#include <stdint.h>
+
+// define the SVE/SME/SME2 thread status structures
```
```suggestion
// Define the SVE/SME/SME2 thread status structures
```

JDevlieghere commented on lldb/tools/debugserver/source/RNBRemote.cpp:
```diff
@@ -2567,10 +2568,13 @@ rnb_err_t RNBRemote::HandlePacket_QSetProcessEvent(const char *p) {
 }
 
-void register_value_in_hex_fixed_width(std::ostream &ostrm, nub_process_t pid,
+// if a fail_value is provided, a correct-length reply is always provided,
```
```suggestion
// If a fail_value is provided, a correct-length reply is always provided,
```

JDevlieghere commented on lldb/tools/debugserver/source/RNBRemote.cpp:
```diff
@@ -2582,27 +2586,31 @@ void register_value_in_hex_fixed_width(std::ostream &ostrm, nub_process_t pid,
+
+    // Pad out the reply to the correct size to maintain correct offsets,
+    // even if we could not read the register value.
+    std::vector<uint8_t> zeros(reg->nub_info.size, *fail_value);
```
Nit: You call the `vector` zeros assuming that's the fail value, but maybe `failed` would be more generic.

JDevlieghere reviewed (commented):
I did a pass and left some nits but I'm not enough of an expert to review this in more detail. I like the test coverage and I appreciate @DavidSpickett taking the time to review this!

jasonmolenda committed:
Carefully testing SME and non-SME hardware showed some testsuite issues

TestRegisters.py expected there to be no registers unable to be
read.  With an SME system when in non-SME mode, all of the SVE/SME
registers will show as unreadable.

TestGdbRemoteGPacket.py was testing the g/G packet with debugserver
(which lldb doesn't use any more) and this exposed the fact that I
wasn't handling SME/SVE registers correctly in
DNBArchImplArm64::GetRegisterContext/SetRegisterContext.  Fix those
to correctly account for the size of these register contexts (the
ZA register is stored in a vector, instead of having a compile-time
register context size).  Also, RNBRemote::HandlePacket_G() was not
handling the thread specifier properly (it would try to interpret
it as content for the packet, and the hex decoding would fail).
I think this test might be setting the current thread with Hg<tid>
before it sends only "g" or "G", that might work.

In debugserver, changed all stack-allocated DNBRegisterValue objects
to be unique_ptr managed so they're on heap.  This object is now
64k and putting that on the stack could be a problem.  We don't
have multiple DNBRegisterValue objects alive at the same time, so
I didn't dynamically size it to the maximum ZA register size on
the current machine.

Fix the comment formatting suggestions from Jonas.

jasonmolenda commented:
I pushed an update addressing Jonas' suggestions, and also fixing two testsuite issues I found by testing the patch on an SME and non-SME machine running the the same OS.  I'm still seeing one bonus failure in TestFirmwareCorefiles.py on the SME system that I need to debug, but the rest of the testsuite looks good.

jasonmolenda committed:
Small fix to the zeroing of the SME regs before attempting reads

When the process is not in Streaming SVE Mode, all of the SME
thread_get_state calls will fail.  Zero out the buffers that we
were going to write data into, so we don't leave old data in them
that might be sent to lldb (particularly in a GetRegisterContext
"g" packet response, where we need to read all registers to complete
the response).

jasonmolenda committed:
fix whitespace nit

jasonmolenda commented:
I debugged the two test failures I'm seeing on an M4, neither is related to the SME changes.  I will handle those two issues separately, this is ready to land.

JDevlieghere commented on lldb/test/API/commands/register/register/register_command/TestRegisters.py:
```diff
@@ -32,11 +50,19 @@ def test_register_commands(self):
         self.log_enable("registers")
 
+        error_str_matched = False
+        if self.get_sme_available() == True and self.platformIsDarwin():
```
```suggestion
        if self.get_sme_available() and self.platformIsDarwin():
```

JDevlieghere reviewed (approved):
LGTM!

DavidSpickett commented:
I've taken the liberty of repeating the fact that the required APIs are not in a publicly available OS release, at the top of the PR description.

Since we had someone on Discord try this and they could not read the registers. Might save a few enthusiasts some time if they see that note first.

The good thing is that lldb didn't break:
```
Process 24700 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = instruction step over
    frame #0: 0x0000000100003f90 sme`main at main.c:13:5
   10   int main(void) {
   11       __asm("smstart");
   12       __asm("ptrue p0.b");
   13       __asm("fmla z0.s, p0/m, z30.s, z31.s");
-> 14       __asm("smopa za0.s, p0/m, p0/m, z30.b, z30.b");
   15       __asm("luti2  z5.h, zt0, z0[2]");
   16       __asm("smstop");
[...]
# register read -a
Scalable Matrix Extension Registers:
4 registers were unavailable.
```
I think that's exactly what should be happening on an unsupported OS version.

jasonmolenda commented:
> I've taken the liberty of repeating the fact that the required APIs are not in a publicly available OS release, at the top of the PR description.

Thanks, I missed that.  Yeah the unrecognized register flavor will be rejected by the thread_get_state call and we'll report the registers as unavailable on current macOSes.  I've got the macOS version where it is available in the test already so I'll make the comment more explicit.

jasonmolenda committed:
Simplify a conditional in the API test.

jasonmolenda committed:
Merge branch 'main' into add-sme-register-support-to-debugserver

digantdesai commented:
Thanks.
